### ISSUE TITLE
GraphQL - proof of concept

### ISSUE THREAD

**Issue by johndpope** [diaspora/diaspora#8033]
![07256CA9-FD9C-4766-B058-21A277B029DF](https://user-images.githubusercontent.com/289994/59850725-4c95f780-93ae-11e9-8600-130f1a3669d9.png)

There’s a huge demand for graphql technology 

Consider even a GUI  headless version of diaspora just using graphQL.

Need to models / account creation controllers and schema 
https://github.com/rmosolgo/graphql-ruby 

Resources
`stackshare{x}io/graphql`

I have more details /roadmap to share here

If there’s interest I will post

**Comment by johndpope** [comment]
Related https://github.com/diaspora/diaspora/issues/3467#issuecomment-504011394

**Comment by johndpope** [comment]
https://api-uswest.graphcms.com/v1/cjwtn7f200c4w01f44q2xo932/master


{
  __schema {
    types {
      name
    }
  }
}

**Comment by jhass** [comment]
Please. Don't just jump on every hype train. Reality is always much more nuanced.https://blog.graphqleditor.com/the-state-of-graphql-by-reddit/

**Comment by Flaburgan** [comment]
>maybe this could be a ‘graphql proof of concept ‘ and tagged help wanted - and keep it in a feature branch

If you want to hack around and try stuff, feel free, that's what's so awesome about free software projects. But this is not something we're going to put effort into, as we have so few resources that we already struggle to have the basics working.

**Comment by johndpope** [comment]
related - https://github.com/diaspora-orm/plugin-server/issues/1

**Comment by johndpope** [comment]
so this will take a json file and scaffold a graphql + db backend
https://github.com/hasura/graphql-engine/tree/master/community/tools/json2graphql

but this is likely to pay biggest dividends 
https://hasura.io/ - real time graphql on postgres
https://github.com/hasura/graphql-engine

I'm going to investigate booting up diaspora against postgres backend. 
From what I'm reading - hasura comes with nice cli out of the box 
https://docs.hasura.io/1.0/graphql/manual/hasura-cli/index.html

I will be going to lunch soon 12:30 
I created postgres instance on aws - free tier 
doing a bundle install.....

**Comment by johndpope** [comment]
so you can simply follow these instructions https://docs.hasura.io/1.0/graphql/manual/deployment/docker/index.html
configure the docker docker-compose.yaml 
 pass in username + password to end point to postgres server
 **docker-compose up** 

<img width="1591" alt="Screen Shot 2019-06-21 at 3 32 27 pm" src="https://user-images.githubusercontent.com/289994/59899908-ce7f3280-9439-11e9-9316-c7f0105634fc.png">
<img width="1603" alt="Screen Shot 2019-06-21 at 3 32 21 pm" src="https://user-images.githubusercontent.com/289994/59899910-ce7f3280-9439-11e9-9040-fa6ad0c0198d.png">
<img width="1679" alt="Screen Shot 2019-06-21 at 3 30 28 pm" src="https://user-images.githubusercontent.com/289994/59899911-cf17c900-9439-11e9-813b-52f9051aeebc.png">
<img width="1678" alt="Screen Shot 2019-06-21 at 3 30 19 pm" src="https://user-images.githubusercontent.com/289994/59899913-cf17c900-9439-11e9-894c-9da73f818d01.png">

<img width="1540" alt="Screen Shot 2019-06-21 at 3 34 22 pm" src="https://user-images.githubusercontent.com/289994/59899982-10a87400-943a-11e9-8abd-cf75e10335de.png">

<img width="1557" alt="Screen Shot 2019-06-21 at 3 35 44 pm" src="https://user-images.githubusercontent.com/289994/59900050-43eb0300-943a-11e9-949c-6350a4926636.png">



with subscriptions - you should get callbacks when new comments are made on site
<img width="1549" alt="Screen Shot 2019-06-21 at 3 38 10 pm" src="https://user-images.githubusercontent.com/289994/59900181-a17f4f80-943a-11e9-91fc-35bce3541d44.png">

**Comment by johndpope** [comment]
i guess the question is - how to expose the feeds as a custom query. 


<img width="662" alt="Screen Shot 2019-06-21 at 3 44 04 pm" src="https://user-images.githubusercontent.com/289994/59900378-6b8e9b00-943b-11e9-8814-1d7867233ab9.png">

I like how the associated tables are highlighted that need tracking.

**Comment by johndpope** [comment]
so to be clear - the above doesn't impact any code / dockerfile from diaspora it's all separate

but to proceed in getting an alpha version which ties into this codebase 
we could build out a new Dockerfile 

https://medium.com/@cbartling/up-and-running-with-hasura-9b098bec82a7

**docker ps
docker exec -it CONTAINERID /bin/sh**

```
Dockerfile 
FROM hasura/graphql-engine:v1.0.0-beta.2
.... dive in machine and throw a customize config specifically for diaspora

```


To leverage existing SQL queries - we can use 
https://docs.hasura.io/1.0/graphql/manual/queries/custom-functions.html

**Comment by jhass** [comment]
Using the solution, how do you create a new post and ensure it's received by all your contacts? Only a  subset  of those contacts? How do you create a comment and ensure all of the original recipients get it? When  fetching  data, how do you respect  share visibilities? How do you know  whether to return the full profile because  you got a contact  or  a limited one with  only  public data? How do you respect the  public  data  switch in the profile settings? I  could go on forever here...

**Comment by denschub** [comment]
Well, okay, I'll be *that guy*.

Simply exposing our database layer to external access is not a thing we want to do, nor is it a thing that can be done without breaking things, nor is it a thing we are going to support or promote in any shape or form.

The only possible way to implement a GraphQL-like access is to write that layer ourselves, which then - as outlined previously - will actually *increase* the amount of effort needed to build and maintain the API. For the application that is diaspora\*, a REST API as we specified and have done serious engineering progress is the most reasonable way. I also see no reason why we should offer two APIs, because that ain't gonna make our lives easier.

So, to make it short: The answer is no.

**Comment by johndpope** [comment]
So to be clear, only the tables that are tracked are exposed as public, eg people / profiles.
There needs to be some due-diligence in how much is exposed. It’s not everything out of the box.I haven’t looked into openid yet but if they use a header parameter this is easily accommodated. 
There seems to be a misunderstanding that this work would be burdened as something you guys would need to pay and solve.

I already managed to build out a rest api across socialstream ruby on rails  backend Back in 2012 and used it without the front end website layer to power mobile apps.  I understand complexities involved and heavy lifting needed to get over the line. There was over $30k in dev by senior developer which is now moth balled. That codebase I still have access to which may be of interest - it would definitely tick every box you’ve outlined needed for rest v1(apart from openid). Create user follow / read stream / add friend ....the list goes on.

Consider that some support needed for graphql to be provided by hasura - 
The work being done building controllers logic is still valid as long as you’ve separated logic from presentation layer. 


Consider also there are commercial developments underway where people will contribute back to this project. You want to foster that not shut people down. 

There are more goodies under the hood with graphql 
 https://blog.hasura.io/building-a-realtime-chat-app-with-graphql-subscriptions-d68cd33e73f/


All those questions above need answering but short answer would be custom controllers / end points. I did see one company use User / PrivateUser to delineate result subsets.
